src/search/BracketsSearchCursor.js

```diff
@@ -0,0 +1,848 @@
+     * A value of '0' is no match.
+     * A value of '1' is positive match.
+     *
+     * @param   {[[Type]]} levelOfDetail An Array of this size will be created and filled with a pattern matching lines matched
+     * @return {object}   An array filled with line match pattern
+     */
+    SearchCursor.prototype.createMatchedLinePattern = function (levelOfDetail) {
+        this._updateResultsIfNeeded(this);
+        var representationArray = new Uint8Array(levelOfDetail);
+        var docLineIndex = _getdocLineIndex(this.doc);
+        var linesPerArraySlot = docLineIndex.length / levelOfDetail;
+        this.searchResults.fillWithMatchedLinePattern(representationArray);
+        return {linesPerArraySlot: linesPerArraySlot, lineMatchPatternArray: representationArray};
+    };
+
+    /**
+     * Finds the indexes of all matches based on the current search query
+     * The matches can then be navigated and retrieved using the functions of the search cursor.
+     *
+     * @return {number} the count of matches found.
+     */
+    SearchCursor.prototype.scanDocumentAndStoreResultsInCursor = function () {
+        if (_needToIndexDocument(this.doc)) {
+            _indexDocument(this.doc);
+        }
+        var docText = _getDocumentText(this.doc);
+        var docLineIndex = _getdocLineIndex(this.doc);
+        this.searchResults = _createSearchResults(docText, docLineIndex, this.query, this.maxResults, this.currentPosition);
+        this.resultsCurrent = true;
+        return this.getMatchCount();
+    };
+
+
+    /**
+     * Creates a regular expression cursor object that this module will provide to consumers
+     * @return {SearchCursor} A new instance of a search cursor
+     */
+    function _createCursor() {
+        return new SearchCursor();
+    }
+
+    /**
+     * Creates an updatable search cursor which can be used to navigate forward and backward through the results.
+     *
+     * @param {!{document: CodeMirror.Doc, searchQuery: string|RegExp, position: {line: number, ch: number}, ignoreCase: boolean}} properties
+     * @return {Object} The search cursor object
+     */
+    function createSearchCursor(properties) {
+        var searchCursor = _createCursor();
+        searchCursor.setSearchDocumentAndQuery(properties);
+        return searchCursor;
+    }
+
+    /**
+     * Scans the entire document for regular expression matches and calls fnEachMatch for each found match.
+     * Unlike the search cursor, this creates no retained results to be used for back and forward navigation.
+     * This is provided for consumers who wish to leverage the speed provided by the document index, but do
+     * not need to use the features of a cursor.
+     *
+     * @param {!{document: CodeMirror.Doc, searchQuery: string|RegExp, ignoreCase: boolean, fnEachMatch: function({line: number, ch: number}, {line: number, ch: number, match: Array})}} properties
+     */
+    function scanDocumentForMatches(properties) {
```

Comment:
I do not insist on this one, but I always found it rather less readable and more prone to errors. In the case of this specific example, a signature like this would make sense to me:
```javascript
function scanDocumentForMatches(document, query, options)
```
It makes it much more understandable as to what is needed for the function to complete successfully.